<a href="https://colab.research.google.com/github/uzdanska/Machine-Learning/blob/main/supervised/02_regression/01_linear_regression_normal_equation.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### scikit-learn

Strona biblioteki: https://scikit-learn.org

Dokumentacja/User Guide: https://scikit-learn.org/stable/user_guide.html

Podstawowa biblioteka do uczenia maszynowego w języku Python.

Aby zainstalować bibliotekę scikit-learn, użyj polecenia poniżej:
```
!pip install scikit-learn

```
Aby zaktualizować do najnowszej wersji bibliotekę scikit-learn, użyj polecenia poniżej:
```
!pip install --upgrade scikit-learn
```
Kurs stworzony w oparciu o wersję 0.22.1

### Preprocessing danych:


1.   [Import bibliorek](#0)
2.   [Przygotowanie danych](#1)
3.   [Równanie normalne](#3)
4.   [Regresja liniowa przy pomocy scikit-learn](#4)


### <a name="0"></a> 1. Import bibliotek

In [5]:
import numpy as np

### <a name="1"></a> 2. Przygotowanie danych

In [6]:
X1 = np.array([1, 2, 3, 4, 5, 6])
Y = np.array([3000, 3250, 3500, 3750, 4000, 4250])
m = len(X1)

print(f'Lata pracy: {X1}')
print(f'Wynagrodzenie: {Y}')
print(f'Liczba próbek: {m}')

Lata pracy: [1 2 3 4 5 6]
Wynagrodzenie: [3000 3250 3500 3750 4000 4250]
Liczba próbek: 6


Nadajemy wartość m wierszy i 1 kolumna

In [7]:
X1 = X1.reshape(m, 1)
print(X1)
print(X1.shape)

[[1]
 [2]
 [3]
 [4]
 [5]
 [6]]
(6, 1)


Wygenerowanie wektora jedynek

In [8]:
bias = np.ones((m, 1))
print(bias)
print(bias.shape)

[[1.]
 [1.]
 [1.]
 [1.]
 [1.]
 [1.]]
(6, 1)


Dodanie wektora jedynek

In [9]:
X = np.append(bias, X1, axis=1)
print(X)
print(X.shape)

[[1. 1.]
 [1. 2.]
 [1. 3.]
 [1. 4.]
 [1. 5.]
 [1. 6.]]
(6, 2)


### <a class="2"></a> 3. Równanie normalne

Regresja liniowa w $R^2$:


$$ Y = w_0 + w_1X_1$$
$$ Y = W^TX$$


gdzie:

$$ W = \left[\begin{array}{ccc}
w_0\\
w_1\\
\end{array}\right]$$

$$ X = \left[\begin{array}{ccc}
1\\
X_1\\
\end{array}\right]$$

stąd:

$$W^T = [w_0, w_1]$$

$$ Y = W^TX = [w_0, w_1] * \left[\begin{array}{ccc}
1\\
X_1\\
\end{array}\right] = w_0 + w_1X_1$$


Równanie normalne - równanie pozwalające obliczyć minimum funkcji straty (o ile istnieje)
$$ W = (X^TX)^{-1}(X^TY)$$

In [10]:
X

array([[1., 1.],
       [1., 2.],
       [1., 3.],
       [1., 4.],
       [1., 5.],
       [1., 6.]])

In [11]:
Y

array([3000, 3250, 3500, 3750, 4000, 4250])

In [16]:
W = np.dot(np.linalg.inv(np.dot(X.T, X)), np.dot(X.T, Y))
W

array([2750.,  250.])

Podział na kroki:

1. $$ X^TX$$



In [17]:
XT_X = np.dot(X.T, X)
XT_X

array([[ 6., 21.],
       [21., 91.]])

2. $$ (X^TX)^{-1}$$

In [23]:
TRANSPOSE_XT_X = np.linalg.inv(XT_X)
TRANSPOSE_XT_X

array([[ 0.86666667, -0.2       ],
       [-0.2       ,  0.05714286]])

3. $$ (X^TY)$$

In [21]:
XT_Y = np.dot(X.T, Y)
XT_Y

array([21750., 80500.])

4. $$ W = (X^TX)^{-1}(X^TY)$$

In [24]:
W = np.dot(TRANSPOSE_XT_X, XT_Y)
W

array([2750.,  250.])

Końcowa postąc modelu:

$$ Y = 2750 + 250 X_1$$

### <a class="3"></a> 4. Regresja liniowa przy pomocy scikit-learn


intercept_ - współczynnik w0, miejsce przecięcia z osią Y
ceof_[0] - współczynnik w1

In [26]:
from sklearn.linear_model import LinearRegression

regression = LinearRegression()
regression.fit(X1, Y)

print(regression.intercept_)
print(regression.coef_[0])

2750.0
250.0
